src/equations/bbm_1d.jl

```diff
@@ -0,0 +1,196 @@
+    invImD2 = lu(I - sparse(solver.D2))
+    eta2 = zeros(RealT, nnodes(mesh))
+    eta2_x = zero(eta2)
+    eta_x = zero(eta2)
+    etaeta_x = zero(eta2)
+    eta_xx = zero(eta2)
+    cache = (; invImD2, eta2, eta2_x, eta_x, etaeta_x, eta_xx, solver.D1, solver.D2)
+    if solver.D1 isa PeriodicUpwindOperators
+        eta_x_upwind = zero(eta2)
+        cache = (; cache..., eta_x_upwind)
+    end
+    return cache
+end
+
+# Discretization that conserves the mass for eta and the modified energy for periodic boundary conditions, see
+# - Hendrik Ranocha, Dimitrios Mitsotakis and David I. Ketcheson (2020)
+#   A Broad Class of Conservative Numerical Methods for Dispersive Wave Equations
+#   [DOI: 10.4208/cicp.OA-2020-0119](https://doi.org/10.4208/cicp.OA-2020-0119)
+function rhs!(dq, q, t, mesh, equations::BBMEquation1D, initial_condition,
+              ::BoundaryConditionPeriodic, source_terms, solver, cache)
+    (; invImD2, eta2, eta2_x, eta_x, etaeta_x) = cache
+    if solver.D1 isa PeriodicUpwindOperators
+        (; eta_x_upwind) = cache
+    end
+
+    eta, = q.x
+    deta, = dq.x
+    # energy and mass conservative semidiscretization
+    @trixi_timeit timer() "hyperbolic" begin
+        @.. eta2 = eta^2
+        if solver.D1 isa PeriodicDerivativeOperator ||
+           solver.D1 isa UniformPeriodicCoupledOperator ||
+           solver.D1 isa FourierDerivativeOperator
+            mul!(eta2_x, solver.D1, eta2)
+            mul!(eta_x, solver.D1, eta)
+            @.. etaeta_x = eta * eta_x
+            # split form
+            @.. deta = -(1 / 3 * (eta2_x + etaeta_x) + eta_x)
+        elseif solver.D1 isa PeriodicUpwindOperators
+            mul!(eta2_x, solver.D1.central, eta2)
+            mul!(eta_x_upwind, solver.D1.minus, eta)
+            mul!(eta_x, solver.D1.central, eta)
+            @.. etaeta_x = eta * eta_x
+            # split form
+            @.. deta = -(1 / 3 * (eta2_x + etaeta_x) + eta_x_upwind)
+        else
+            @error "unknown type of first-derivative operator: $(typeof(solver.D1))"
+        end
+    end
+
+    @trixi_timeit timer() "source terms" calc_sources!(dq, q, t, source_terms, equations,
+                                                       solver)
+
+    @trixi_timeit timer() "elliptic" begin
+        solve_system_matrix!(deta, invImD2, equations)
+    end
+    return nothing
+end
+
+"""
+    energy_total_modified!(e, q, equations::BBMEquation1D, cache)
+
+Return the modified total energy `e` of the primitive variables `q_global` for the
+[`BBMEquation1D`](@ref). The `energy_total_modified`
+is a conserved quantity (for periodic boundary conditions).
+
+It is given by
+```math
+\\frac{1}{2} \\eta^2 + \\frac{1}{2} \\eta_{xx}.
+```
+
+`q_global` is a vector of the primitive variables at ALL nodes.
+`cache` needs to hold the SBP operators used by the `solver`.
+
+See also [`energy_total_modified`](@ref).
+"""
+function energy_total_modified!(e, q_global, equations::BBMEquation1D, cache)
```

Comment:
I used the `energy_total_modified` for this function because it uses derivatives of $\eta$. Does this make sense in your eyes? We don't have an `energy` for the `BBMEquation1D` up to now.

Reply:
For BBM, the quadratic invariant should be $\eta^2 + \eta_x^2$, not $\eta^2 + \eta_{xx}$? In periodic domains, you can write this as $\eta (\eta - \eta_{xx})$, but it has to be quadratic

Reply:
I am not sure what we could use as a reasonable "energy" of the system - since the energy we have used so far has always been the shallow water energy. But BBM does not have the shallow water system as a subsystem, so I don't know whether there is a canonical "energy" we can use here

Reply:
> For BBM, the quadratic invariant should be η 2 + η x 2 , not η 2 + η x x ? In periodic domains, you can write this as η ( η − η x x ) , but it has to be quadratic

You're right. I was too fast. Let me fix that quickly.

Reply:
> I am not sure what we could use as a reasonable "energy" of the system - since the energy we have used so far has always been the shallow water energy. But BBM does not have the shallow water system as a subsystem, so I don't know whether there is a canonical "energy" we can use here

Ok, so any idea of a good name for this invariant then?

Reply:
>> For BBM, the quadratic invariant should be η 2 + η x 2 , not η 2 + η x x ? In periodic domains, you can write this as η ( η − η x x ) , but it has to be quadratic

> You're right. I was too fast. Let me fix that quickly.


I just pushed d7f07eaca0722b581aba151a0fb59d7be5357bdc. I hope this fixes this.

Reply:
> > I am not sure what we could use as a reasonable "energy" of the system - since the energy we have used so far has always been the shallow water energy. But BBM does not have the shallow water system as a subsystem, so I don't know whether there is a canonical "energy" we can use here

> Ok, so any idea of a good name for this invariant then?

Or do you think `energy_total_modified` (and `entropy_modified`) is fine?

Reply:
`energy_total_modified` is fine, I guess. It is in some sense consistent with what we have so far. But we should document this explicitly in the docstrings (of the equations).

Reply:
Ok, I agree.

Reply:
> Would you recommend implementing eta * (eta - D2 * eta) instead?

Yes 👍 That's the "correct" way to get this invariant

Reply:
I switched to `eta * (eta - D2 * eta)` in [673a72d](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/pull/150/commits/673a72d7cc7463a527082dba94508417672aeddc).